main/src/library/MutList.flix

```diff
@@ -801,7 +835,9 @@ namespace MutList {
     ///
     /// Modifying `l` while using an iterator has undefined behavior and is dangerous.
     ///
-    pub def iterator(l: MutList[a, r]): Iterator[a, r] \ Write(r) =
+    /// Andreas: "Function reads from len, so I have added Read(r) effect.
+    ///
+    pub def iterator(l: MutList[a, r]): Iterator[a, r] \ { Read(r), Write(r) } =
```

Comment:
Question 1: What does the interface of `Iterator` look like? 

Question 2: Given that we can create an iterator over MutList, what prevents re-entrancy by exploiting the iterator? Is that a property of (a) `MutList`, (b) `def iterator()`, (c) both, (d) something else?

Reply:
1. Looking at `Iterator`, I am not sure _why_ the `new` method needs to have any effect on the region `r` at all. 
2. (a) `def iterator()` breaks the abstraction, so not really on `MutList` at this point.  
    (b) This function has `Write(r)` effect - there might be a discussion on if we should allocate `i` in the same region. Outside of that I can be convinced the iterator somewhat guards us from re-entrancy.